### PR TITLE
Only reload cells that changed for MessagesTableViewController #190

### PR DESCRIPTION
**Pull request by iglushchuk**
This Pull Request fixes issue #190 

Here I added two different dataSources: one for iOS 13 using "Diffable data source" API and for previous iOS versions I made data source, which reloads with tableView.reloadData(). Messages storage moves to datasources, so now they are source of truth.

I was thinking about making a protocol, but that protocol should be generic, and I would have to use type erasure on this protocol, which looked like overengineering and I would loose type safety, provided by generic MessagesTableViewController. Also I didn't figure out how to unit test this data sources.

Looking forward to code review :)

### CODE REVIEW

File: CriticalMass/MessagesDefaultDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    func tableView(_: UITableView, numberOfRowsInSection _: Int) -> Int {
+        messages.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(ofType: T.self)
+        cell.setup(for: messages[indexPath.row])
+
+        return cell
+    }
```
Comment (lennet) [inline CriticalMass/MessagesDefaultDataSource.swift:36]: Maybe we can move that code to MessagesDataSource to share it with MessageDiffableDataSource

File: CriticalMass/ChatMessage.swift [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 import Foundation
 
-public struct ChatMessage: Codable, Equatable {
+public struct ChatMessage: Codable, Equatable, Hashable {
```
Comment (lennet) [inline CriticalMass/ChatMessage.swift:10]:
```suggestion
public struct ChatMessage: Codable, Hashable {
```

**Comment by lennet** [comment]
> 
> [...]Also I didn't figure out how to unit test this data sources. [...]
> 

We could unit test by creating a pure tableview in the test case and attach it to the datasource and send some messages to the datasource and see if the row count and the content matches and updates. That would be nice but not super important 

Maybe a SnapshotTests with both DataSource would be nice just to see that everything works since we won't regularly test the default data source.

File: CriticalMass/MessagesTableViewController.swift [hunk 3]
```diff
@@ -70,30 +53,16 @@ class MessagesTableViewController<T: IBConstructableMessageTableViewCell>: UITab
...
-        self.messages = messages
-    }
-
-    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        let cell = tableView.dequeueReusableCell(ofType: T.self)
-        cell.setup(for: messages[indexPath.row])
-        return cell
+        dataSource.messages = messages
+        updateNoMessageCountIfNeeded()
+        tableView.reloadData()
```
[the remark this answers is not in the document]
Comment (iglushchuk) [reply CriticalMass/MessagesTableViewController.swift:59]: You are right, I will fix that

File: CriticalMass/MessagesDefaultDataSource.swift (hunk 1 above)
Comment (iglushchuk) [reply CriticalMass/MessagesDefaultDataSource.swift:36]: Good idea, I'll do that

**Comment by iglushchuk** [comment]
Fixed issues and added unit/snapshot tests, check it please 😃

**Comment by iglushchuk** [comment]
I saw that animation on tweets screen was not that good, so I turned off animation in diffable if we populate tableview form zero to N cells. In other cases there should be animation.

File: CriticalMapsSnapshotTests/MessagesTableViewControllerSnapshotTests.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
+//
```
Comment (lennet) [inline CriticalMapsSnapshotTests/MessagesTableViewControllerSnapshotTests.swift:1]: Nice Idea to use Snapshot tests here 👏

**Pull request merged**